File: backend/app/api/schemas/prospector.py
from __future__ import annotations

import re

from pydantic import BaseModel, Field, field_validator


class ProspectorManualRequest(BaseModel):

    cnpj: str = Field(..., description="CNPJ com ou sem máscara")
    site: str | None = Field(None, description="URL do site (opcional)")

    instagram: str | None = Field(None, description="@handle ou URL")
    facebook: str | None = Field(None, description="handle ou URL")
    linkedin: str | None = Field(
        None, description="LinkedIn do decisor (pessoa, não empresa)"
    )
    email: str | None = None
    telefone: str | None = None
    whatsapp: str | None = None

    @field_validator("cnpj")
    @classmethod
    def _cnpj_so_digitos_validos(cls, v: str) -> str:
        if not v:
            raise ValueError("CNPJ é obrigatório")
        digits = "".join(c for c in v if c.isdigit())
        if len(digits) != 14:
            raise ValueError(
                f"CNPJ deve ter 14 dígitos, veio com {len(digits)}"
            )
        return digits

    @field_validator("site")
    @classmethod
    def _site_opcional_mas_se_vier_valido(cls, v: str | None) -> str | None:
        if v is None:
            return None
        v = v.strip()
        if not v:
            return None
        if not re.match(r"^(https?://)?[\w\-.]+\.[a-z]{2,}", v, re.IGNORECASE):
            raise ValueError(f"URL inválida: {v!r}")
        return v



class ContatoOut(BaseModel):
    nome: str | None = None
    cargo: str | None = None
    email: str | None = None
    telefone: str | None = None
    whatsapp: str | None = None
    linkedin: str | None = None
    decisor: bool = False


class EmpresaOut(BaseModel):
    nome: str | None = None
    razao_social: str | None = None
    cnpj: str | None = None
    cidade: str | None = None
    estado: str | None = None
    setor: str | None = None
    tamanho: str | None = None
    capital_social: float | None = None
    site: str | None = None
    instagram: str | None = None
    facebook: str | None = None
    notas: str | None = None
    notion_page_id: str | None = None


class LeadOut(BaseModel):
    empresa: EmpresaOut
    contatos: list[ContatoOut] = Field(default_factory=list)


class ProspectorPreviewResponse(BaseModel):

    success: bool = True
    fonte_cnpj: str | None = Field(
        None, description="Qual fonte resolveu o CNPJ: 'brasilapi' ou 'opencnpj'"
    )
    lead: LeadOut


class ProspectorRunResponse(BaseModel):

    success: bool = True
    fonte_cnpj: str | None = None
    lead: LeadOut
    notion_empresa_id: str | None = None
    notion_contatos_ids: list[str] = Field(default_factory=list)


class LeadHistoryItem(BaseModel):

    empresa_nome: str
    cnpj: str | None = None
    cidade: str | None = None
    estado: str | None = None
    setor: str | None = None
    qtd_contatos: int = 0
    notion_empresa_id: str | None = None
    enviado_em: str | None = Field(
        None, description="ISO 8601 — extraído do nome do arquivo"
    )
    arquivo: str = Field(..., description="Nome do arquivo de backup local")


class LeadHistoryResponse(BaseModel):
    items: list[LeadHistoryItem]
    total: int


class ApiError(BaseModel):
    success: bool = False
    error: str
    detail: str | None = None

File: frontend/lib/types/prospector.ts
// Prospector — coleta/enriquecimento de leads (CNPJ, contatos, histórico).

export interface ProspectorManualRequest {
  cnpj: string;
  site?: string | null;
  instagram?: string | null;
  facebook?: string | null;
  linkedin?: string | null;
  email?: string | null;
  telefone?: string | null;
  whatsapp?: string | null;
}

export interface Contato {
  nome: string | null;
  cargo: string | null;
  email: string | null;
  telefone: string | null;
  whatsapp: string | null;
  linkedin: string | null;
  decisor: boolean;
}

export interface Empresa {
  nome: string | null;
  razao_social: string | null;
  cnpj: string | null;
  cidade: string | null;
  estado: string | null;
  setor: string | null;
  tamanho: string | null;
  capital_social: number | null;
  site: string | null;
  instagram: string | null;
  facebook: string | null;
  notas: string | null;
  notion_page_id: string | null;
}

export interface Lead {
  empresa: Empresa;
  contatos: Contato[];
}

export type FonteCnpj = 'brasilapi' | 'opencnpj' | null;

export interface ProspectorPreviewResponse {
  success: boolean;
  fonte_cnpj: FonteCnpj;
  lead: Lead;
}

export interface ProspectorRunResponse {
  success: boolean;
  fonte_cnpj: FonteCnpj;
  lead: Lead;
  notion_empresa_id: string | null;
  notion_contatos_ids: string[];
}

export interface LeadHistoryItem {
  empresa_nome: string;
  cnpj: string | null;
  cidade: string | null;
  estado: string | null;
  setor: string | null;
  qtd_contatos: number;
  notion_empresa_id: string | null;
  enviado_em: string | null; // ISO 8601
  arquivo: string;
}

export interface LeadHistoryResponse {
  items: LeadHistoryItem[];
  total: number;
}
